eth_abi/registry.py

```diff
@@ -0,0 +1,131 @@
+from cytoolz import curry
+
+from .utils.parsing import collapse_type
+
+
+class ABIRegistryError(Exception):
+    pass
+
+
+class NoEntryFound(ABIRegistryError):
+    pass
+
+
+class PredicateMapping:
+    """
+    Acts as a mapping from predicate functions to values.  Values are retrieved when their
+    corresponding predicate matches a set of given inputs.  Predicates can also be labeled to
+    facilitate removal from the mapping.
+    """
+    def __init__(self):
+        self._values = {}
+        self._labeled_predicates = {}
+
+    def add(self, predicate, value, label=None):
+        if predicate in self._values:
+            raise ValueError('Predicate already exists')
+
+        if label is not None:
+            if label in self._labeled_predicates:
+                raise ValueError('Predicate with label "{}" already exists'.format(label))
+
+            self._labeled_predicates[label] = predicate
+
+        self._values[predicate] = value
+
+    def find(self, *args, **kwargs):
+        results = tuple(
+            value for predicate, value in self._values.items()
+            if predicate(*args, **kwargs)
+        )
+
+        if len(results) == 0:
+            raise ValueError('No matching values found')
+
+        if len(results) > 1:
+            raise ValueError('Multiple matching values found')
+
+        return results[0]
+
+    def remove_by_reference(self, predicate):
+        # Delete the predicate mapping to the previously stored value
+        try:
+            del self._values[predicate]
+        except KeyError:
+            raise KeyError('Predicate not found')
+
+        # Delete any label which refers to this predicate
+        try:
+            label = self._label_for_predicate(predicate)
+        except ValueError:
+            pass
+        else:
+            del self._labeled_predicates[label]
+
+    def _label_for_predicate(self, predicate):
+        for key, value in self._labeled_predicates.items():
+            if value is predicate:
+                return key
```

Comment:
Does this data structure contain an *implicit* guarantee that this loop either matches 1 or 0 values and cannot match multiple predicates?  If so, it might be worth adding a comment here to ensure future readers of this code know that case is not something that was *overlooked*.

Reply:
It does give that guarantee since both keys and values in that dict must be unique (or an exception is raised).  I'll add a comment to that effect.